---
title: "CTDquerier: Case study on Environmental Chemicals and Human Non-Communicable Diseases"
author:
- name: Carles Hernandez-Ferrer
  affiliation: ISGlobal, Centre for Research in Environmental Epidemiology (CREAL)
- name: Juan R. Gonzalez
  affiliation: ISGlobal, Centre for Research in Environmental Epidemiology (CREAL)
  email: juanr.gonzalez@isglobal.org
date: "`r doc_date()`"
package: "`r pkg_ver('CTDquerier')`"
csl: biomed-central.csl
bibliography: case_study.bib
vignette: >
  %\VignetteIndexEntry{CTDquerier: Case study on Environmental Chemicals and Human Non-Communicable Diseases}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document2:
    toc_float: true
---

# Introduction

## Health Impacts of Chemicals

The health impact of chemicals is determined by a process of assessment which aims to provide a consensus scientific description of the risks of chemical exposures. The International Programme on Chemical Safety (IPCS) of the World Health Organization (WHO) is one of the organizations that works to establish the scientific basis for the sound management of chemicals, and to strengthen national capabilities and capacities for chemical safety. In its responsibilities are the publication of the reports and other related documents so that governments and international and national organizations can use them as the basis for taking preventive actions against adverse health and environmental impacts.

The leading cause of deaths and disability worldwide is ischaemic heart disease. The second largest contributor to global mortality is stroke. The 35% of the first and the 42% of the second could be prevented by reducing or removing exposure to chemicals such as from  ambient air pollution, household air pollution, second-hand smoke and lead [@WHOHealthEnvironment2016].

Chemicals such as heavy metals, pesticides, solvents, paints, detergents, kerosene, carbon monoxide and drugs lead to unintentional poisonings at home and in workplace. This type of unintentional poisonings are estimated to cause 193 000 deaths annually [@WHOHealthEnvironment2016][@WHOKnownsUnknowns2016].

It is well known that air pollution and second-hand smoke are risk factors for adverse pregnancy outcomes like low birth weight, prematurity and stillbirths [@Khader2011]. Antenatal exposure to second-hand smoke for example was estimated to increase the overall risk for stillbirths by 23% and for congenital malformations by 13%. There are, furthermore, potential links between various chemicals and adverse pregnancy outcomes or congenital malformations, though evidence is limited [@WHOHealthEnvironment2016].

Moreover, second-hand smoke and air pollution can lead to the development of asthma. Air pollution additionally provokes asthma exacerbations and increases related hospital admissions [@WHOHealthEnvironment2016][@Macintyre2013].


## The Comparative Toxicogenomics Database

The Comparative Toxicogenomics Database (*CTDbase*; http://ctdbase.org) is a public resource for toxicogenomic information manually curated from the peer-reviewed scientific literature, providing key information about the interactions of environmental chemicals with gene products and their effect on human disease [@CTDbase2003][@CTDbase2017].

*CTDbase* is curated by professional biocurators who leverage controlled vocabularies, ontologies and structured notation to code a triad of core interactions describing chemical-gene, chemical-disease and gene-disease relationships, which are then internally integrated to generate inferred chemical-gene-disease networks [@CTDbase2017][@CTDbase2011].

*CTDbase* if offered to public by using a web-based interface that includes basic and advanced query options to access data for sequences, references, and toxic agents, and a platform for analysing sequences.

*CTDbase* includes more than 30.5 million toxicogenomic connections relating chemicals/drugs, genes/proteins, diseases, exposures, Gene Ontology (GO) annotations, pathways (KEGG/Reactome), and gene interaction modules [@CTDbase2017].


## `CTDquerier` R package

`CTDquerier` is an R package that allows to R users to download basic data from *CTDbase* about genes, chemicals and diseases. Once the user's input is validated allows to query *CTDbase* to download the information of the given input from the other modules. Tables \@ref(tab:CTDquerier-queries) indicates the information `CTDquerier` can download from *CTDbase* depending on the user's input.

|           | Gene Interaction | Disease | Pathway | Gene Ontology | Chemical Interaction |
|:----------|:----------------:|:-------:|:-------:|:-------------:|:--------------------:|
|Genes      | X                | X       | X       | X             | X                    |
|Chemical   | X                | X       | X       | X             |                      |
|Disease    | X                |         | X       |               | X                    |

: (\#tab:CTDquerier-queries) Relation between input accepted by `CTDquerier` and the data it is capable to download from *CTDbase*.

`CTDquerier` can be installed using `devtools`. To install `CTDquerier` run the following command in an R session:

```{r eval=FALSE}
devtools::install_github( "carleshf/CTDquerier" )
```

# Case-Study

The case-study is based on the results shows in the article entitled "Case-control admixture mapping in Latino populations enriches for known asthma-associated genes", from *Torgerson et. al.* [@Torgerson2012]. The genes from **Table E1** were collected and used as starting point for this document.

## Introduction

Asthma is a common long term inflammatory disease of the airways of the lungs. It is characterized by variable and recurring symptoms, reversible airflow obstruction, and bronchospasm. Symptoms include episodes of wheezing, coughing, chest tightness, and shortness of breath. Asthma is thought to be caused by a combination of genetic and environmental factors. Environmental factors include exposure to air pollution and allergens. Diagnosis is usually based on the pattern of symptoms, response to therapy over time, and spirometry. Asthma is classified according to the frequency of symptoms, forced expiratory volume in one second, and peak expiratory flow rate. It may also be classified as atopic or non-atopic where atopy refers to a predisposition toward developing a type 1 hypersensitivity reaction.

The Genetics of Asthma in Latino Americans (GALA) study [@Price2009] includes subjects with asthma and their biological parents recruited from schools, clinics, and hospitals at 4 sites: San Francisco Bay Area, New York City, Puerto Rico, and Mexico City. Patients were contacted to participate in the study if approved by their primary physician.  On the basis of interviews and questionnaire data, subjects were included in the study if they were between the ages of 8 and 40 years with physician-diagnosed mild to moderate-to-severe asthma and had experienced 2 or more symptoms in the previous 2 years at the time of recruitment (including wheezing, coughing, and/or shortness of breath) and if both parents and 4 sets of grandparents self-identified as being of either Puerto Rican or Mexican ethnicity. 

Genotyping of GALA subjects was performed by using the Affymetrix 6.0 GeneChip Array that contained more than 900,000 SNPs before quality control measures. Markers were filtered based on 95% call rates, Hardy-Weinberg equilibrium P values of greater than $10^{-6}$. Subjects were filtered based on 95% call rates and consistency between genetic and reported sex. SNPs were filtered for linkage disequilibrium (LD) based on $r^2$ values of greater than 0.5 before performing a principal component analysis.

After quality control, the total number of SNPs included in the study was 729,685, and the total number of subjects was 529 children with asthma (253 Mexican and 276 Puerto Rican subjects) and 347 control subjects (158 Mexican and 189 Puerto Rican subjects). 

Comparisons of local ancestry between cases and control subjects identified 62 admixture mapping peaks with a difference in local ancestry at a P value of less than $10^{-3}$. A total of 29 peaks were identified in Puerto Rican subjects and 33 in Mexican subjects. We observed a significant enrichment of previously identified asthma-associated genes within the 62 admixture mapping peaks (P = .0051). Six (9.6%) of the 62 admixture mapping peaks contained at least 1 gene that had been positively associated with asthma, as compiled from the Genetic Association Database [@Becker2004]. No individual SNP within the 62 admixture mapping peaks was significantly associated with asthma after multiple testing correction, including SNPs imputed from the 1000 Genomes Project. 

## Set-Up

### Reading Data File

The **Table E1** from the GALA Study [@Torgerson2012] was downloaded and converted into a `.csv` file. Then it can be loaded with function `read.csv`.

```{r gala_read_csv}
table_e1_csv <- system.file(
  paste0( "extdata", .Platform$file.sep, "gala_table_e1.csv" ), 
  package="CTDquerier"
)
table_e1 <- read.csv( table_e1_csv, stringsAsFactors = FALSE )
```

From this table, we are interested in the last column, called *Genes*. We will filter this column, removing the `NA` values.

```{r gala_remove_na_file}
dim( table_e1 )
table_e1 <- table_e1[ table_e1$Genes != "NA ", ]
dim( table_e1 )
```
We filtered 17 rows that had no genes. No we will extract the genes from each one of the remaining rows.

```{r gala_create_list}
gala_genes <- trimws( unlist( strsplit( table_e1$Genes, "," ) ) )
length( gala_genes )
```

We end up with a total list of 305 related to asthma genes.

### Querying *CTDbase* by Genes

The first step to use `CTDquery` is to use the generated list of genes of interest to query *CTDabse*. This is done using the function `query_ctd_gene`, that takes a list of genes to validate it to *CTDbase gene-vocabulary* and then performs a query per gene in order to download their associated information.

```{r load_ctdquerier, message=FALSE}
library( CTDquerier )
```
```{r gale_query, eval=FALSE}
gala <- query_ctd_gene( terms = gala_genes, verbose = TRUE )
```
Since this process can be hight time-consuming, `CTDquerier` has the result of the query and it can be loaded with:

```{r gala_data}
data( gala, package = "CTDquerier" )
gala
```

## Questions answered about GALA genes

### How many of GALA genes are in *CTDbase*?

Original genes from Table E1 are stored in the variable `gala_genes`. The result obtained from querying *CTDbase* with those genes is stored in the variable `gala`.

The list `gala_genes` has a length has 305 genes but `gala` was generated using 258 terms. This means that 47 genes from the GALA study were not present in *CTDbase*. This can be plotted using the method `plot` of a `CTDquery` object.

```{r gala_plot_query, message=FALSE}
library( ggplot2 )
plot( gala ) + ggtitle( "Lost & Found Genes from GALA Study" )
```

The names of the *lost* genes from GALA study can be obtained using the method `get_terms`. The names follow:

```{r gala_lost}
get_terms( gala )[[ "lost" ]]
```

### How many diseases are associated to GALA genes according to *CTDbase*?

The call to `query_ctd_gene` done in previous section has download "all the information available" in *CTDbase* for the given GALA genes. The line with `#Diseases` that can be seen when an object of class `CTDquery` is shown indicates the number of relationships between a gene or chemical and diseases.

```{r gala_show_2}
gala
```

For the GALA genes, there are a total of 278'761 associations between genes and diseases. The table with the associations between genes and diseases can be obtained using the method `extract` from `psygenet2r` R package.

```{r gala_gda_all}
library( psygenet2r )
gala_all_diseases <- extract( gala, index_name = "diseases" )
colnames( gala_all_diseases )
dim( gala_all_diseases )
```

We can corroborate that the table with the associations between gene and diseases has close to ~300K rows. We can also check that the genes used to create this table:

```{r gala_disease_genes}
length( unique( gala_all_diseases$GeneSymbol ) )
sum( get_terms( gala )[[ "found" ]] %in% 
    unique( gala_all_diseases$GeneSymbol ) )
sum( !get_terms( gala )[[ "found" ]] %in% 
    unique( gala_all_diseases$GeneSymbol ) )
```

Then, 246 genes were used to create the table of associations between genes and diseases, and 12 have not been used. We can check the name of the genes that do not contributed to create the table of associations between genes and diseases:

```{r gala_diseases_no_genes}
get_terms( gala )[[ "found" ]][ 
    !get_terms( gala )[[ "found" ]] %in% unique( gala_all_diseases$GeneSymbol )
]
```

The number of unique diseases associated to the GALA genes is:

```{r gala_diseases_unique}
length( unique( gala_all_diseases$Disease.Name ) )
```

### What is the level of evidence of the association between GALA genes and asthma according to *CTDbase*?

The first step to locate the level of evidence of the association between GALA genes and asthma is to locate asthma in the table of associations between genes and diseases.

```{r gala_diseases_asthma_names}
grep( "Asthma", unique( gala_all_diseases$Disease.Name ), value = TRUE )
```

The previous code shows the five entries existing in the table of associations between GALA genes and *CTDbase* diseases.

Then, to look for the level of evidence of the association between GALA genes and asthma we will subset the table `gala_all_diseases` selecting only the entries for `"Asthma"`.

```{r gala_diseases_asthma}
gala_asthma <- gala_all_diseases[ 
    gala_all_diseases$Disease.Name == "Asthma" , 
]
dim( gala_asthma )
```

The subsetting resulted in a table with 209 entries. *CTDbase* has two ways to compute the level of evidence of the association between a gene and a disease. Both of them can be obtained from the table `gala_asthma` and are the columns `Inference.Score` and `Reference.Count`. The first contains the *CTDbase* inference-score of gene-disease association in base of a series of criterion defined by *CTDbase*'s developers while the second are just the number of bibliographical references enforcing the relation.

Then, the mean of the *Inference Score* can be interpreted as the level of evidence of the association between GALA genes and asthma according to *CTDbase*:

```{r gala_diseases_asthma_evidence}
mean( gala_asthma$Inference.Score, na.rm = TRUE )
```

Otherwise one could say that GALA genes have a total reference coud of association with asthma of 2'982.

```{r gala_diseases_asthma_reference}
sum( gala_asthma$Reference.Count, na.rm = TRUE )
```

Then we can check for the *Inference Score* of the association of each gene with a single diseases, in the case Asthma. We can obtain this score by plotting the *Inference Score* with GALA genes and sub-setting the diseases by Asthma:

```{r gala_diseases_asthma_evidence_plot}
plot( gala, index_name = "disease", subset.y = "Asthma", filter.score = 20 ) +
    ggtitle( "Evidence of the association between GALA genes and Asthma" )
```

### Which genes are associated to asthma according to *CTDbase*?

The way we can obtain all the information related to a disease from *CTDbase* is near the same as the way we got the information from genes. The function to be used is `query_ctd_dise`:

```{r ctd_asthma}
asthma <- query_ctd_dise( terms = "Asthma", verbose = TRUE )
```

As we can see from the log obtained when `verbose` is set to `TRUE`, the information downloaded from *CTDabse* by `CTDquerier` for diseases is composed by disease-gene interactions, disease-chemical interactions and pathway interactions (KEGG).

```{r ctd_asthma_show}
asthma
```

So, to know the exact number of genes associated with Asthma according to *CTDbase* we extract the proper table and count the unique genes in it.

```{r ctd_asthma_n_genes}
ctd_asthma <- extract( asthma, index_name = "gene interactions" )
length( unique( ctd_asthma$Gene.Symbol ) )
```

So, according to *CTDbase*, there are 22'704 to Asthma. But are we sure we are talking about **only** Asthma?

```{r ctd_asthma_table, results="asis"}
library(knitr)
tt <- as.data.frame( table( ctd_asthma$Disease.Name ) )
colnames( tt ) <- c( "Disease", "Frequency" )
kable( tt[ order( tt$Frequency, decreasing = TRUE ), ] )
```

Hence, the *true* answer is 22'258, not 22'704.

### Which chemicals are associated to GALA genes according to *CTDbase*?

The object `gala` has a total of 16'611 gene-chemical interaction registers. We can obtain this table using the `extract` method.

```{r gala_chemicals}
gala_chem <- extract( gala, index_name = "chemical interactions" )
colnames( gala_chem )
length( unique( gala_chem$Chemical.Name ) )
```

So, GALA genes are *associated* with 1'829 unique chemicals according to *CTDabse*. The `Reference.Count` of these *associations* can be plotted. First, lets see its distribution:

```{r, gala_chemicals_table, results="asis"}
kable( t( table( gala_chem$Reference.Count ) ) )
```

The plot for this information corresponds to a heat-map. The arguments `subset.x` and `subset.y` are used to filter the elements in X-axis (genes) and Y-axis (chemicals). The argument `filter.score` allows to filter by `Reference.Count`'s value.

```{r, gala_chemicals_plot}
plot( gala, index_name = "chemical", filter.score = 6 )
```

### Which chemicals are associated to asthma according to *CTDbase*?

To anwer this question we only need to follow the steps done in the previous section.

```{r ctd_asthma_chem}
ctd_asthma_chem <- extract( asthma, index_name = "chemical interactions" )
colnames( ctd_asthma_chem )
length( unique( ctd_asthma_chem$Chemical.Name ) )
```

Hence, the short answer is 5'282. Then we can go for the `Inference.Score` or `Reference.Count` plot.

```{r ctd_asthma_plot}
plot( asthma, index_name = "chemical", subset.x = "Asthma", filter.score = 30 ) +
    ggtitle( "Evidence of the association between GALA genes and chemicals" )
```

### Which chemicals are associated to both GALA genes and to asthma according to *CTDbase*?

The table `gala_chem` contains the relation between GALA genes and the chemicals associated to these genes according to *CTDbase*. The table `ctd_asthma_chem` contains the chemicals associated to Asthma according to *CTDbase*.

To find the chemicals that are both related to GALA genes and Asthma, according to *CTDbase* we apply an intersection to both tables.

```{r intersect_gala_asthma_chem_1}
intr_chem <- intersect( gala_chem$Chemical.Name, ctd_asthma_chem$Chemical.Name )
length( intr_chem )
```
So, there are 1'485 chemicals shearing association between GALA genes and Asthma.

```{r intersect_gala_asthma_chem_2}
length( intr_chem ) / nrow( gala_chem ) * 100
length( intr_chem ) / nrow( ctd_asthma_chem ) * 100
```
In other words, the 8.94% of the total association from GALA genes to *CTDbase* chemicals are shared with the Asthma related chemicals. Also that the 15.77% of the chemicals related to Asthma according to *CTDbase* are shared with the chemicals associated with GALA genes.

The following code has the aim to create a `data.frame` (called `dta`) with only three columns: the chemical's name and their reference count in both GALA and Asthma chemicals sets.

```{r intersect_gala_asthma_chem_plot_1}
gala_chem_r <- gala_chem[ gala_chem$Chemical.Name %in% intr_chem, ]
gala_chem_r <- gala_chem_r[ !duplicated( gala_chem_r$Chemical.Name ), ]
ctd_asthma_chem_r <- ctd_asthma_chem[ ctd_asthma_chem$Chemical.Name %in% intr_chem, ]
ctd_asthma_chem_r <- ctd_asthma_chem_r[ !duplicated( ctd_asthma_chem_r$Chemical.Name ), ]

dta <- merge(
    gala_chem_r[ , c( "Chemical.Name", "Reference.Count" ) ],
    ctd_asthma_chem_r[ , c( "Chemical.Name", "Reference.Count" ) ],
    by = "Chemical.Name"
)
colnames( dta ) <- c( "Chemical.Name", "Reference.Gala", "Reference.Asthma" )
dta <- dta[ 
    order( dta$Reference.Gala, dta$Reference.Asthma, decreasing = TRUE ), 
]
dta[1:3, ]
```

Having this `data.frame`, we can use the function `leaf_plot` to compare the number of references for the top 25 chemicals.

```{r intersect_gala_asthma_chem_plot_2}
leaf_plot( dta[1:25, ], label = "Chemical.Name", 
    valueLeft = "Reference.Gala", valueRight = "Reference.Asthma",
    titleLeft = "GALA", titleRight = "Asthma"
)
```

### Are the GALA genes significantly associated with asthma according to *CTDabse*?

In other to see if GALA genes are significantly associated with asthma we will use the method `enrich`. this method allows to perform a Fisher test given two objects of class \code{CTDquery}.

```{r gala_enrich_asthma}
enrich( gala, asthma )
```

In this case, `enrich` is taking the genes in the `gala` object, the genes in the `asthma` object and all the genes in *CTDbase* and performing a Fisher test with these numbers (see `fisher.test`).

### Are the GALA genes significantly enriched in Air Pollutants chemicals according to *CTDabse*?

First of all we need to obtain the information about air pollutants from *CTDbase*. This is done using the function `query_ctd_chem`.

```{r air_ctd}
air <- query_ctd_chem( terms = "Air Pollutants" )
air
```
Once we obtained all the information about air pollutants from *CTDbase* we can use the already seen method `enrich` to see if the GALA genes are enriched in air Pollutants' chemicals.

```{r gala_enrich_air}
enrich( gala, air )
```

In this case, `enrich` is using the chemicals in `gala` object, the chemicals in `air` object and all the chemicals in *CTDbase* to perform the Fisher test.

# Session Info.

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# Bibliography
